File: app/src/main/java/com/cradle/neptune/database/RoomDatabaseManager.java
package com.cradle.neptune.database;

import android.content.Context;
import android.os.AsyncTask;

import com.cradle.neptune.model.Reading;
import com.cradle.neptune.model.ReadingManager;
import com.cradle.neptune.utilitiles.GsonUtil;
import com.cradle.neptune.utilitiles.Util;

import org.threeten.bp.ZonedDateTime;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class RoomDatabaseManager implements ReadingManager, HealthFacilityManager {
    private CradleDatabase cradleDatabase;

    public RoomDatabaseManager(CradleDatabase cradleDatabase) {
        this.cradleDatabase = cradleDatabase;
    }

    @Override
    public void addNewReading(Context context, Reading reading) {
        // add the new reading into the DB.
        reading.dateLastSaved = ZonedDateTime.now();
        if (reading.readingId == null || reading.readingId.equals("") ||
                reading.readingId.toLowerCase().equals("null")) {
            reading.readingId = UUID.randomUUID().toString();
        }

        ReadingEntity readingEntity = new ReadingEntity(reading.readingId,
                reading.patient.patientId, GsonUtil.getJson(reading), reading.isUploaded());
        cradleDatabase.readingDaoAccess().insertReading(readingEntity);

        // update all the other patient's records .. because thats the way it is...
        List<Reading> readings = getReadings(context);
        for (Reading r : readings) {
            if (r.patient.patientId.equals(reading.patient.patientId) && r.readingId != reading.readingId) {
                if (r.isNeedRecheckVitals()) {
                    r.dateRecheckVitalsNeeded = null;
                    updateReading(context, r);
                }
            }
        }

    }

    @Override
    public void updateReading(Context context, Reading reading) {
        reading.dateLastSaved = ZonedDateTime.now();
        ReadingEntity readingEntity = new ReadingEntity(reading.readingId,
                reading.patient.patientId, GsonUtil.getJson(reading), reading.isUploaded());
        cradleDatabase.readingDaoAccess().update(readingEntity);
    }

    @Override
    public List<Reading> getReadings(Context context) {
        return new GetAllReadingsAsyncTask(cradleDatabase, false, "").doInBackground();
    }

    @Override
    public Reading getReadingById(Context context, String id) {
        ReadingEntity readingEntity = cradleDatabase.readingDaoAccess().getReadingById(id);
        if (readingEntity == null) {
            return null;
        }
        Reading r = GsonUtil.makeObjectFromJson(readingEntity.getReadDataJsonString(), Reading.class);
        String patientId = r.patient.patientId;
        Util.ensure(readingEntity.getPatientId() == patientId ||
                patientId.equals(r.patient.patientId));
        return r;

    }

    @Override
    public void deleteReadingById(Context context, String readingID) {
        ReadingEntity readingEntity = cradleDatabase.readingDaoAccess().getReadingById(readingID);
        if (readingEntity != null) {
            cradleDatabase.readingDaoAccess().delete(readingEntity);
        }

    }

    @Override
    public void deleteAllData(Context context) {
        cradleDatabase.readingDaoAccess().deleteAllReading();
    }

    @Override
    public List<Reading> getReadingByPatientID(Context context, String patientId) {
        return new GetAllReadingsAsyncTask(cradleDatabase, true, patientId).doInBackground();
    }

    @Override
    public void addAllReadings(Context context, List<Reading> readings) {
        List<ReadingEntity> readingEntities = new ArrayList<>();
        for (Reading reading : readings) {
            ReadingEntity readingEntity = new ReadingEntity(reading.readingId,
                    reading.patient.patientId, GsonUtil.getJson(reading), reading.isUploaded());
            readingEntities.add(readingEntity);
        }
        cradleDatabase.readingDaoAccess().insertAll(readingEntities);
    }

    @Override
    public List<Reading> getUnuploadedReadings() {
        List<ReadingEntity> readingEntities = cradleDatabase.readingDaoAccess().getAllUnUploadedReading();
        List<Reading> readings = new ArrayList<>();
        for (ReadingEntity readingEntity : readingEntities) {
            Reading r = GsonUtil.makeObjectFromJson(readingEntity.getReadDataJsonString(), Reading.class);
            readings.add(r);
        }
        return readings;
    }

    /**
     * since we dont want to block the main UI thread, we have to create a seperate thread for large queries.
     */
    private class GetAllReadingsAsyncTask extends AsyncTask<Void, Void, List<Reading>> {
        WeakReference<CradleDatabase> readingEntitiesDatabaseWeakReference;
        private boolean readingByPatientId = false;
        private String patientId = "";

        GetAllReadingsAsyncTask(CradleDatabase cradleDatabase, boolean readingByPatientId, String patientId) {
            this.readingEntitiesDatabaseWeakReference = new WeakReference<>(cradleDatabase);
            this.readingByPatientId = readingByPatientId;
            this.patientId = patientId;
        }

        @Override
        protected List<Reading> doInBackground(Void... url) {
            List<Reading> readings = new ArrayList<>();
            List<ReadingEntity> readingEntities;
            if (readingByPatientId) {
                readingEntities = readingEntitiesDatabaseWeakReference.get().readingDaoAccess().getAllReadingByPatientId(patientId);
            } else {
                readingEntities = readingEntitiesDatabaseWeakReference.get().readingDaoAccess().getAllReadingEntities();
            }
            for (ReadingEntity readingEntity : readingEntities) {
                Reading r = GsonUtil.makeObjectFromJson(readingEntity.getReadDataJsonString(), Reading.class);
                readings.add(r);
                String patientId = r.patient.patientId;
                Util.ensure(readingEntity.getPatientId() == patientId ||
                        patientId.equals(r.patient.patientId));
            }
            return readings;
        }
    }

    /**
     * HealthFacility functions
     */

    @Override
    public void insert(HealthFacilityEntity healthFacilityEntity) {
        cradleDatabase.healthFacilityDaoAccess().insert(healthFacilityEntity);
    }

    @Override
    public void removeFacilityById(String id) {
        HealthFacilityEntity healthFacilityEntity =
                cradleDatabase.healthFacilityDaoAccess().getHealthFacilityById(id);
        cradleDatabase.healthFacilityDaoAccess().delete(healthFacilityEntity);
    }

    @Override
    public void insertAll(List<HealthFacilityEntity> healthCareFacilityEntities) {
        cradleDatabase.healthFacilityDaoAccess().insertAll(healthCareFacilityEntities);
    }

    @Override
    public List<HealthFacilityEntity> getAllFacilities() {
        return cradleDatabase.healthFacilityDaoAccess().getAllHealthFacilities();
    }

    @Override
    public HealthFacilityEntity getFacilityById(String id) {
        return cradleDatabase.healthFacilityDaoAccess().getHealthFacilityById(id);
    }

    @Override
    public List<HealthFacilityEntity> getUserSelectedFacilities() {
        return cradleDatabase.healthFacilityDaoAccess().getAllUserSelectedHealthFacilities();
    }
}

File: app/src/main/java/com/cradle/neptune/model/RoomDatabaseManager.kt
package com.cradle.neptune.model

import android.content.Context
import com.cradle.neptune.database.CradleDatabase
import com.cradle.neptune.database.HealthFacilityEntity
import org.threeten.bp.ZonedDateTime
import java.util.UUID

class RoomDatabaseManager(val database: CradleDatabase) : ReadingManager {

    private val readingDao get() = database.readingDaoAccess()
    private val healthFacilityDao get() = database.healthFacilityDaoAccess()

    /**
     * Adds a new reading to the database.
     */
    override fun addNewReading(context: Context, reading: Reading) {
        reading.metadata.dateLastSaved = ZonedDateTime.now()

        // Since this is the first time we are adding this reading to the
        // database it probably does not have an id so we'll give it one.
        val entity = reading.toEntity()
        readingDao.insertReading(entity)

        // Update other readings associated with this patient.
        for (r in getReadingByPatientID(context, reading.patientId)) {
            if (r.isVitalRecheckRequired) {
                r.dateRecheckVitalsNeeded = null
                updateReading(context, r)
            }
        }
    }

    override fun updateReading(context: Context, reading: Reading) {
        reading.metadata.dateLastSaved = ZonedDateTime.now()

        // We assume that the reading already exists in the database and, as
        // such, has an id. If it doesn't then that is a programmer error and
        // we crash with a null pointer exception.
        val entity = reading.toEntity()
        readingDao.update(entity)
    }

    /**
     * Retrieves all readings from the database.
     */
    override fun getReadings(context: Context): List<Reading> {
        TODO("Not yet implemented")
    }

    /**
     * Returns the reading with a given [id].
     */
    override fun getReadingById(context: Context, id: String): Reading? {
        return getReadingById(id)
    }

    private fun getReadingById(id: String): Reading? {
        val entity = readingDao.getReadingById(id) ?: return null
        val jsonString = entity.readDataJsonString!!
        val json = JsonObject(jsonString)
        return Reading.unmarshal(json)
    }

    /**
     * Constructs a [RetestGroup] for [reading] by grouping together all
     * related readings.
     */
    override fun getRetestGroup(reading: Reading): RetestGroup {
        val readings = mutableListOf<Reading>()
        readings.addAll(reading.previousReadingIds.mapNotNull { id -> this.getReadingById(id) })
        readings.add(reading)
        return RetestGroup(readings)
    }

    /**
     * Deletes the reading with a given [id].
     *
     * If no such reading exists in the database, then this method does nothing.
     */
    override fun deleteReadingById(context: Context, id: String) {
        val entity = readingDao.getReadingById(id) ?: return
        readingDao.delete(entity)
    }

    /**
     * Deletes all readings from the database.
     */
    override fun deleteAllData(context: Context) {
        readingDao.deleteAllReading()
    }

    // TODO: Rename to getReadingsByPatientId to be consistent
    override fun getReadingByPatientID(context: Context, id: String): List<Reading> {
        TODO("Not yet implemented")
    }

    /**
     * Inserts all supplied readings into the database.
     *
     * If any of the readings do not have identifiers, new ones are generated
     * for them before they are inserted into the database.
     */
    override fun addAllReadings(context: Context, readings: List<Reading>) =
        readings.map(Reading::toEntity).forEach(readingDao::insertReading)

    override fun getUnuploadedReadings(): List<Reading> = readingDao.allUnUploadedReading
        .map {
            // The entity's json string field may not be `null`, the only
            // reason it is declared optional is to conform to the legacy
            // database schema.
            val json = JsonObject(it.readDataJsonString!!)
            Reading.unmarshal(json)
        }

    override fun insert(healthFacilityEntity: HealthFacilityEntity) =
        healthFacilityDao.insert(healthFacilityEntity)

    override fun removeFacilityById(id: String) {
        val facility = healthFacilityDao.getHealthFacilityById(id) ?: return
        healthFacilityDao.delete(facility)
    }

    override fun insertAll(facilities: List<HealthFacilityEntity>) =
        healthFacilityDao.insertAll(facilities)

    override fun getAllFacilities() = healthFacilityDao.allHealthFacilities

    override fun getFacilityById(id: String) = healthFacilityDao.getHealthFacilityById(id)

    override fun getUserSelectedFacilities() = healthFacilityDao.allUserSelectedHealthFacilities

    override fun updateFacility(healthFacilityEntity: HealthFacilityEntity) =
        healthFacilityDao.update(healthFacilityEntity)
}
